### PR TITLE
fix: Serialize custom fields on linked entities

### PR DESCRIPTION
**Pull request by marikaner**
Fix the serialization on linked entities. Related to #849.
When doing this I realized, that we have some issues with linked entities, making it hard to fix this without changing some internal handling of remote state and serialization. I ended up with an optional parameter for the serialization. The `diff` parameter indicates whether the whole entity or just the changed properties should be serialized.

Background: with this it is possible to serialize the entity using some functionality that the entity provides and serializing the diff only, by simply skipping the unchanged properties. Before we already passed the diff to this function, but this diff was no entity (although the type told us so - I fixed that). Therefore at the point where I wanted to serialize links the entity class information was lost.

### CODE REVIEW

File: packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts [hunk 1]
```diff
@@ -47,6 +47,8 @@ describe('CreateRequestBuilderV4', () => {
     testPostRequestOutcome(actual, entity.setOrInitializeRemoteState());
   });
 
+  // TODO: This test is quite useless and wrong. It does not fail even if we remove the multilink.
```
Comment (marikaner) [inline packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts:50]: This is not a nice comment, we should remove it ;)

Comment (marikaner) [reply packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts:50]: @jjtang1985 I am not sure what to do with this file. To be honest it seems to me that all the tests are quite useless. We are basically checking that the deserializer works. Which is already checked in the deserialize

File: packages/core/src/odata-v2/request-builder/update-request-builder.spec.ts [hunk 2]
```diff
@@ -290,12 +290,11 @@ describe('UpdateRequestBuilder', () => {
     expect(actual['remoteState']).toEqual(entity);
   });
 
-  it('warns if navigaton properties are sent', async () => {
+  it.only('warns if navigaton properties are sent', async () => {
```
Comment (jjtang1985) [inline packages/core/src/odata-v2/request-builder/update-request-builder.spec.ts:293]:
```suggestion
  it('warns if navigaton properties are sent', async () => {
```

File: packages/core/src/odata-v2/entity-serializer.spec.ts [hunk 3]
```diff
@@ -111,43 +129,23 @@ describe('entity-serializer', () => {
...
-      ComplexTypeProperty: {
-        StringProperty: complexTypeProp.stringProperty,
-        BooleanProperty: complexTypeProp.booleanProperty,
-        Int16Property: complexTypeProp.int16Property
-      },
-      StringProperty: stringProperty
-    });
-  });
-
-  it('should serialize an entity with custom field', () => {
```
Comment (jjtang1985) [inline packages/core/src/odata-v2/entity-serializer.spec.ts:137]: I personally would keep these two tests.

File: packages/core/src/odata-v2/entity.spec.ts [hunk 4]
```diff
@@ -1,114 +0,0 @@
-import { TestEntity } from '../../test/test-util/test-services/v2/test-service';
```
Comment (jjtang1985) [inline packages/core/src/odata-v2/entity.spec.ts:1]: I assume you move the tests from here to the common.

File: packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts (hunk 1 above)
Comment (jjtang1985) [reply packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts:50]: Would you fix this test or make a follow-up? 😄

File: packages/core/src/odata-v4/entity.spec.ts [hunk 5]
```diff
@@ -55,8 +55,8 @@ describe('entity v4', () => {
       })
       .build();
   });
-  it('getCurrentMapKeys', () => {
-    const currentState = entity.getCurrentMapKeys();
+  it('asObject', () => {
```
Comment (jjtang1985) [inline packages/core/src/odata-v4/entity.spec.ts:58]: in general, `asObject` is not a v4 specific function. Should we keep it in the common or v2 folder?

File: packages/core/src/odata-v4/entity-serializer.spec.ts [hunk 6]
```diff
@@ -117,4 +118,39 @@ describe('entity-serializer', () => {
...
+      )
+      .build();
+    expect(serializeEntityV4(testEntity, TestEntity)).toEqual({
+      to_SingleLink: {
+        custom: 'custom'
+      }
+    });
+  });
+
+  it('should consider only changed items when diff is true', () => {
```
Comment (jjtang1985) [inline packages/core/src/odata-v4/entity-serializer.spec.ts:137]: These are not v4 specific, right? Could we move them to v2/common then?

**Review by jjtang1985 — CHANGES_REQUESTED** [review]
Looks good, please consider my comments.

File: packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts (hunk 1 above)
Comment (marikaner) [reply packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts:50]: The question is - what should be tested here? The v4 create request builder doesn't do anything, therefore I would say it is too trivial to be tested.

File: packages/core/src/odata-v2/entity-serializer.spec.ts (hunk 3 above)
Comment (marikaner) [reply packages/core/src/odata-v2/entity-serializer.spec.ts:137]: I moved them to common.

File: packages/core/src/odata-v2/entity.spec.ts (hunk 4 above)
Comment (marikaner) [reply packages/core/src/odata-v2/entity.spec.ts:1]: Yes.

File: packages/core/src/odata-v4/entity.spec.ts (hunk 5 above)
Comment (marikaner) [reply packages/core/src/odata-v4/entity.spec.ts:58]: Yes, common makes more sense. Unfortunately there are so many untidy things around this, that I wasn't sure what to touch and what not... But I can move this test.

Comment (marikaner) [reply packages/core/src/odata-v4/entity.spec.ts:58]: Now, I remembered, why I didn't move it. There are clashes with some of the other tests and I want this to be done thoughtfully. I will create a ticket for this.

File: packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts (hunk 1 above)
Comment (marikaner) [reply packages/core/src/odata-v4/request-builder/create-request-builder.spec.ts:50]: I now removed the test file, but also created a follow up ticket to ensure our odata tests are complete and correct ;)

File: packages/core/src/odata-v4/entity-serializer.spec.ts (hunk 6 above)
Comment (marikaner) [reply packages/core/src/odata-v4/entity-serializer.spec.ts:137]: Also created follow up for this, as I think this is a bigger task, if we want to do it right ;)

**Review by jjtang1985 — APPROVED** [review]

**Pull request merged**